### ISSUE TITLE
Operational attributes raise LDAPCursorError in abstraction layer

### ISSUE THREAD

**Issue by styopa** [cannatag/ldap3#311]
Hi Giovanni,

Thanks for your awesome library! It's super helpful and reliable. There's a bug I noticed, however, when using the abstraction layer and requesting operational attributes. ObjectDef (just as expected) is not aware of the server's operational attributes (such as **memberOf**, **entryUUID**, or **pwdChangedTime** in case of OpenLDAP). So when the search returns such attributes, it causes an error:
`ldap3.core.exceptions.LDAPCursorError: attribute 'entryUUID' not in object class 'inetOrgPerson, posixAccount, shadowAccount, inetLocalMailRecipient, ldapPublicKey' for entry uid=dummy,ou=people,dc=example,dc=org`

I'm using version 2.2.0, but the code is the same in master, in ldap3/abstract/cursor.py:83-93.
[oper.py.txt](https://github.com/cannatag/ldap3/files/822545/oper.py.txt)

**Comment by cannatag** [comment]
Hi, the problem seems related to operational attributes that are not returned in the schema. The LDAP server should return the syntax of these attributes in the schema, but they're missing.

Will investigate to find a suitable solution for using the in the abstraction layer.

**Comment by styopa** [comment]
It seems like we can't get them from schema at all, they're internal use after all. Maybe it'll make sense to just store unknown attribs in some ciDict instead of raising an exception, when operational attribs are requested?

**Comment by cannatag** [comment]
Anyway you can tell ldap3 to not check for a specific attribute. Try this:

```
from ldap3 import get_config_parameter, set_config_parameter
missing_attrs = ['memberOf', 'entryUUID', 'pwdChangedTime']  # all the missing attributes you need

attrs = get_config_parameter('ATTRIBUTES_EXCLUDED_FROM_CHECK')
attrs.extend(missing_attrs)
set_config_parameter('ATTRIBUTES_EXCLUDED_FROM_CHECK', attrs)
```
then add the requested attributes to the object definition:

```
oc_person = ["inetOrgPerson", "posixAccount", "shadowAccount", "inetLocalMailRecipient", "ldapPublicKey"]
person = ObjectDef(object_class = oc_person, schema = conn)

person += missing_attrs  # this creates the missing AttrDef in the ObjectDef
```

then when you query the cursor you'll get back the missing attributes too.

Let me know if this is good for you.

Bye,
Giovanni

**Comment by cannatag** [comment]
Did you try the proposed solution?

**Comment by styopa** [comment]
Thanks, Giovanni! Unfortunately, the suggested solution will not work in my case. It would require hardcoding values for each LDAP server implementation, which I'm trying to avoid.

In my opinion, operational attributes are meant to be "unexpected" by the client. So if the client requests attributes that are unknown to it at this point, it should be prepared to receive something unknown at validation time. My idea is: when operational attributes are requested, don't raise the exception, but store the raw attributes.

 Please go ahead and close this issue, I'll suggest a patch later.

**Comment by cannatag** [comment]
Operational attribute should be exposed in the schema. Anyway ldap3 has an 'offline schema' feature that let you instruct the Server object on the schema when it's not received by the LDAP server. Maybe this can help you.

Giovannj

**Comment by cannatag** [comment]
Fix will be released in 2.2.3

**Comment by cannatag** [comment]
Released in ldap3 v. 2.2.3.
